### PR TITLE
Bullets (with shooting animation and sound)

### PR DESCRIPTION
**Pull request by Alanine42**
wooh, another milestone done!

- left click to shoot. 
- the server will only let you shoot if it's more than 3 ticks since your last shot. 
- third-person shooting
    - the camera (plus forwardDirection) casts a ray forward, anything it hits is your ideal target
    - your gun casts another ray toward that ideal target, whatever it hits along the way is the real target. 

biu biu biu ---> x _ x

- shooting resolution is handled by `BulletVsPlayerHandler`. it deals damage (unless you shot the egg), swaps positions if the target is ~~dead~~, etc.
- the server notifies clients about all `BulletTrail (shooterID, startPos, endPos)`s spawned during the game tick.  
---
- clients render each bullet trail with a long quad (rectangle). the quad moves from your gun to the target in 5 frames.
- shooting animation is played *from the beginning* each time a player requests to shoot (`ClientGame::updateAnimations`). 
- shooting audio is played each time your bullet is shot (`ClientGame::handleBulletPacket`).

### CODE REVIEW

**Comment by Alanine42** [comment]
(why force push) I squashed the last 7 commits together. they were trial and error (particularly with rendering the stupid crosshair that wouldn't show the right pixels). some bugs were discovered occasionally during gameplay and chasing them down was ~~confusing~~ fun. 

now things *should be* in a good state and bug-free. check it out and enjoy some shooting

-------
New things added since the last comment:
- crosshair 
- thicker bullet trail

File: client/src/ClientGame.cpp [hunk 1]
```diff
@@ -19,14 +19,18 @@ ClientGame::ClientGame()
 void ClientGame::updateAnimations(std::bitset<NUM_STATES> movementEntityStates[]) {
     for (unsigned int i = 0; i < NUM_PLAYER_ENTITIES; i++) {
         unsigned int movementIndex = playerIndices[i];
-        if (movementEntityStates[movementIndex][MOVING_HORIZONTALLY]) {
+        if (movementEntityStates[i][MOVING_HORIZONTALLY]) {
```
Comment (Melody-Ruth) [inline client/src/ClientGame.cpp:22]: Why did you switch to using i here? Did something change about how movementEntityStates works? The point of movementIndex = playerIndices[i] was to go from an index into the players (0, 1, 2, or 3) to an index into the general list of moving entities. I know we do put the players at 0, 1, 2, and 3 in the movement list as well, but I felt like we were less likely to introduce bugs this way and would extend better to doing similar things to other categories of entities.

File: server/src/bge/System.cpp [hunk 2]
```diff
@@ -199,9 +192,10 @@ namespace bge {
             forwardDirection.y = 0;
             forwardDirection.z = sin(glm::radians(req.yaw));
             forwardDirection = glm::normalize(forwardDirection);
-            req.forwardDirection = forwardDirection;
+            req.forwardDirection = glm::normalize(forwardDirection);
```
Comment (Melody-Ruth) [inline server/src/bge/System.cpp:195]: Isn't this already normalized on the previous line?

File: server/src/bge/System.cpp [hunk 3]
```diff
@@ -199,9 +192,10 @@ namespace bge {
...
 
-            glm::vec3 rightwardDirection = glm::normalize(glm::cross(forwardDirection, glm::vec3(0, 1, 0)));
+            glm::vec3 rightwardDirection = glm::cross(forwardDirection, glm::vec3(0, 1, 0));
```
Comment (Melody-Ruth) [inline server/src/bge/System.cpp:197]: Good catch, I didn't notice that forwardDirection and (0,1,0) are orthogonal so we don't need to do the extra normalization.

**Review by Metarus — APPROVED** [review]
couple things melody noted but apart from that, lgtm! really like the double raycast for shooting direction, seems very logically handled and should lead to fluid gameplay

File: client/src/ClientGame.cpp (hunk 1 above)
Comment (Alanine42) [reply client/src/ClientGame.cpp:22]: ur right!

once upon time there was a bug that you can't see other player's shooting animation. i thought this was the cause (but its not). 

I'll be using playerIndices from now on, it sets a good ground to extend other types of moving entities

File: server/src/bge/System.cpp (hunk 2 above)
Comment (Alanine42) [reply server/src/bge/System.cpp:195]: oh shoot, i must've been blind, or was git merge behaving again?

good catch, fixed

**Review by Melody-Ruth — APPROVED** [review]
Lgtm! It runs fine on my machine. Lots of great work here, and I'm excited for the shooting feature!

**Review by AnnLe4869 — APPROVED** [review]
Work well on my end

**Review by realmatthewpeng — APPROVED** [review]
Great stuff!

**Pull request merged**